### PR TITLE
[klippa] serializer: Added packed_map hash table and pop_pack()

### PR DESCRIPTION
**Pull request by qxliu76**

### CODE REVIEW

File: klippa/src/serialize.rs [hunk 1]
```diff
@@ -165,6 +178,12 @@ impl Serializer {
         !!self.errors
     }
 
+    pub(crate) fn only_overflow(&self) -> bool {
+        self.errors == SerializeErrorFlags::SERIALIZE_ERROR_OFFSET_OVERFLOW
+            || self.errors == SerializeErrorFlags::SERIALIZE_ERROR_INT_OVERFLOW
+            || self.errors == SerializeErrorFlags::SERIALIZE_ERROR_ARRAY_OVERFLOW
+    }
```
Comment (cmyr) [inline klippa/src/serialize.rs:185]: iirc this check is returning `false` if the bitset contains some combination of these flags?

File: klippa/src/serialize.rs [hunk 2]
```diff
@@ -188,6 +207,106 @@ impl Serializer {
...
+            let hash_old = hash_one_pool_idx(to.0, &self.data, &self.object_pool);
+            self.packed_map
+                .del(to.0, hash_old, &self.data, &self.object_pool);
+        }
+
+        {
+            let to_obj = self.object_pool.get_obj_mut(to.0).unwrap();
+            to_obj.virtual_links.append(&mut from_vec);
+        }
+
```
Comment (cmyr) [inline klippa/src/serialize.rs:304]: I don't think these braces are doing anything? (this still compiles fine without them)

File: klippa/src/serialize.rs [hunk 3]
```diff
@@ -188,6 +207,106 @@ impl Serializer {
...
+            self.head = obj.head;
+            self.tail -= len;
+
+            obj_head = obj.head;
+            obj.head = self.tail;
+            obj.tail = self.tail + len;
+
+            //TODO: consider zerocopy here
+            self.data.copy_within(obj_head..obj_head + len, self.tail);
+        }
```
Comment (cmyr) [inline klippa/src/serialize.rs:248]: These braces don't seem necessary? (maybe they were in an earlier iteration?)

**Review by cmyr — APPROVED** [review]
Just a quick pass looking at rust stuff; maybe @garretrieger can take a look as well, since he is more familiar with the harfbuzz code?

File: klippa/src/serialize.rs (hunk 2 above)
Comment (qxliu76) [reply klippa/src/serialize.rs:304]: Oh yeah it seems these braces are not needed with mem::take. Thank you!

File: klippa/src/serialize.rs [hunk 4]
```diff
@@ -70,20 +76,27 @@ struct Object {
     next_obj: Option<PoolIdx>,
 }
 
+impl Object {
+    fn fini(&mut self) {
```
[the remark this answers is not in the document]
Comment (garretrieger) [reply klippa/src/serialize.rs:80]: or maybe reset()?

File: klippa/src/serialize.rs [hunk 5]
```diff
@@ -188,6 +207,106 @@ impl Serializer {
...
+            obj.tail = self.tail + len;
+
+            //TODO: consider zerocopy here
+            self.data.copy_within(obj_head..obj_head + len, self.tail);
+        }
+
+        let mut hash = 0_u64;
+        // introduce flags to avoid mutable borrow and immutable borrow issues
+        let mut obj_duplicate = false;
+        let mut dup_obj_idxes = (0, 0);
```
Comment (garretrieger) [inline klippa/src/serialize.rs:253]: obj_duplicate and dup_obj_idxes could be combined into a single Option<...> field

File: klippa/src/serialize.rs [hunk 6]
```diff
@@ -188,6 +207,106 @@ impl Serializer {
...
+                obj_duplicate = true;
+                dup_obj_idxes = *entry;
+            };
+        }
+
+        if obj_duplicate {
+            self.merge_virtual_links(pool_idx, dup_obj_idxes);
+            let obj = self.object_pool.get_obj_mut(pool_idx).unwrap();
+            obj.fini();
+            self.object_pool.release(pool_idx);
```
Comment (garretrieger) [inline klippa/src/serialize.rs:269]: could we have release automatically call .fini()?

File: klippa/src/serialize.rs [hunk 7]
```diff
@@ -351,4 +551,171 @@ mod test {
...
+        let n = UfWord::new(10);
+        assert_eq!(s.embed(n), Ok(3));
+        assert_eq!(s.head, 5);
+        assert_eq!(s.pop_pack(true), Some(2));
+        assert_eq!(s.packed_map.hash_table.len(), 2);
+        //check to make sure head is rewinded
+        assert_eq!(s.head, 3);
+        //check that tail is updated
+        assert_eq!(s.tail, 95);
+    }
```
Comment (garretrieger) [inline klippa/src/serialize.rs:720]: Some additional tests to add:
- pop_pack(false) when a duplicate object exists
- de-duplication with virtual links (ie. checking merge virtual links works, and that virtual links don't affect equality for de-duping).
- de-duplication on an object that has real links to test they are being included in hash and equality computations.

File: klippa/src/serialize.rs (hunk 1 above)
Comment (qxliu76) [reply klippa/src/serialize.rs:185]: Good question! It's ported from Harfbuzz. I think these flags are exclusive, maybe @garretrieger can help to confirm?

File: klippa/src/serialize.rs (hunk 3 above)
Comment (qxliu76) [reply klippa/src/serialize.rs:248]: right, no braces are needed here after I moved code around. thank you!

File: klippa/src/serialize.rs (hunk 6 above)
Comment (qxliu76) [reply klippa/src/serialize.rs:269]: good idea! We should do the same thing in harfbuzz as well.

File: klippa/src/serialize.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (qxliu76) [reply klippa/src/serialize.rs]: Done.

File: klippa/src/serialize.rs (hunk 7 above)
Comment (qxliu76) [reply klippa/src/serialize.rs:864]: Thanks! Adding de-duplication test with virtual links uncovered a typo in the code: I used real_links in the merge_virtual_links() API.

File: klippa/src/serialize.rs (hunk 1 above)
Comment (garretrieger) [reply klippa/src/serialize.rs:185]: Not sure why specifically it's that way in hb but unless we have a strong reason otherwise we should just match hb.

**Review by garretrieger — APPROVED** [review]

**Pull request merged**